[Style] Split style and platform transform operation representations

weinig opened the pull request:
#### f87b14ad446e9c8acc2768e313611ac359a6ed33
<pre>
[Style] Split style and platform transform operation representations
<a href="https://bugs.webkit.org/show_bug.cgi?id=299814">https://bugs.webkit.org/show_bug.cgi?id=299814</a>

Reviewed by Antti Koivisto.

In order to remove another remaining use of WebCore::Length, this change
adds a style specific parallel set of types for representing transform
operations to the WebCore::TransformOperation ones.

The new types, deriving from Style::TransformFunction, use Style primitive
numerics for the underlying data, allowing us to replace the WebCore::Length
with Style::LengthPercentage&lt;&gt; in Style::TranslateTransformFunction.

In the pre-existing WebCore::TranslateTransformOperation, WebCore::Length
is replaced by direct usage of float (it might make sense to change this
to double at some point, but for now, float is used to maintain the same
bit accuracy as Length).

When going from the style types to the platform types, via Style::toPlatform,
a FloatSize must be passed in so that Style::LengthPercentage&lt;&gt; in
Style::TranslateTransformFunction can be resolved down to float.

The platform transform types are only used for platform layer animation
needs (both GraphicsLayer and threaded), so going forward, any capabilities
they don&apos;t need can stripped from them. For GraphicsLayer, it may make sense
to remove use of WebCore::TransformOperation entirely, instead having the
ability to go straight from Style::TransformFunction to platform animation
types, be it PlatformCAAnimation or TextureMapperAnimation.

This introduces a bit of duplication of logic for the blending that is not
ideal, so a followup change will aim to add a generic implementation of
the blending algorithms that can be shared.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/BlendingKeyframes.cpp:
* Source/WebCore/animation/KeyframeEffect.cpp:
* Source/WebCore/css/values/CSSValueAggregates.h:
* Source/WebCore/platform/animation/AcceleratedEffect.cpp:
* Source/WebCore/platform/animation/AcceleratedEffect.h:
* Source/WebCore/platform/animation/AcceleratedEffectValues.cpp:
* Source/WebCore/platform/graphics/GraphicsLayer.h:
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/texmap/GraphicsLayerTextureMapper.cpp:
* Source/WebCore/platform/graphics/texmap/GraphicsLayerTextureMapper.h:
* Source/WebCore/platform/graphics/texmap/TextureMapperAnimation.cpp:
* Source/WebCore/platform/graphics/texmap/TextureMapperAnimation.h:
* Source/WebCore/platform/graphics/texmap/coordinated/GraphicsLayerCoordinated.cpp:
* Source/WebCore/platform/graphics/texmap/coordinated/GraphicsLayerCoordinated.h:
* Source/WebCore/platform/graphics/transforms/IdentityTransformOperation.h:
* Source/WebCore/platform/graphics/transforms/Matrix3DTransformOperation.cpp:
* Source/WebCore/platform/graphics/transforms/Matrix3DTransformOperation.h:
* Source/WebCore/platform/graphics/transforms/MatrixTransformOperation.h:
* Source/WebCore/platform/graphics/transforms/PerspectiveTransformOperation.h:
* Source/WebCore/platform/graphics/transforms/RotateTransformOperation.h:
* Source/WebCore/platform/graphics/transforms/ScaleTransformOperation.h:
* Source/WebCore/platform/graphics/transforms/SkewTransformOperation.h:
* Source/WebCore/platform/graphics/transforms/TransformOperation.h:
* Source/WebCore/platform/graphics/transforms/TransformOperations.cpp:
* Source/WebCore/platform/graphics/transforms/TransformOperations.h:
* Source/WebCore/platform/graphics/transforms/TransformOperationsSharedPrimitivesPrefix.h:
* Source/WebCore/platform/graphics/transforms/TranslateTransformOperation.cpp:
* Source/WebCore/platform/graphics/transforms/TranslateTransformOperation.h:
* Source/WebCore/rendering/RenderLayer.cpp:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
* Source/WebCore/rendering/TextAutoSizing.cpp:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+CSSValueConversion.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+CSSValueCreation.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Conversions.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Serialization.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes.h:
* Source/WebCore/style/values/transforms/StyleRotate.cpp:
* Source/WebCore/style/values/transforms/StyleRotate.h:
* Source/WebCore/style/values/transforms/StyleScale.cpp:
* Source/WebCore/style/values/transforms/StyleScale.h:
* Source/WebCore/style/values/transforms/StyleTransform.cpp:
* Source/WebCore/style/values/transforms/StyleTransform.h:
* Source/WebCore/style/values/transforms/StyleTransformFunction.cpp:
* Source/WebCore/style/values/transforms/StyleTransformFunction.h:
* Source/WebCore/style/values/transforms/StyleTransformList.cpp:
* Source/WebCore/style/values/transforms/StyleTransformList.h:
* Source/WebCore/style/values/transforms/StyleTranslate.cpp:
* Source/WebCore/style/values/transforms/StyleTranslate.h:
* Source/WebCore/style/values/transforms/functions/StyleMatrix3DTransformFunction.cpp: Added.
* Source/WebCore/style/values/transforms/functions/StyleMatrix3DTransformFunction.h: Added.
* Source/WebCore/style/values/transforms/functions/StyleMatrixTransformFunction.cpp: Added.
* Source/WebCore/style/values/transforms/functions/StyleMatrixTransformFunction.h: Added.
* Source/WebCore/style/values/transforms/functions/StylePerspectiveTransformFunction.cpp: Added.
* Source/WebCore/style/values/transforms/functions/StylePerspectiveTransformFunction.h: Added.
* Source/WebCore/style/values/transforms/functions/StyleRotateTransformFunction.cpp: Added.
* Source/WebCore/style/values/transforms/functions/StyleRotateTransformFunction.h: Added.
* Source/WebCore/style/values/transforms/functions/StyleScaleTransformFunction.cpp: Added.
* Source/WebCore/style/values/transforms/functions/StyleScaleTransformFunction.h: Added.
* Source/WebCore/style/values/transforms/functions/StyleSkewTransformFunction.cpp: Added.
* Source/WebCore/style/values/transforms/functions/StyleSkewTransformFunction.h: Added.
* Source/WebCore/style/values/transforms/functions/StyleTransformFunctionBase.cpp: Added.
* Source/WebCore/style/values/transforms/functions/StyleTransformFunctionBase.h: Added.
* Source/WebCore/style/values/transforms/functions/StyleTransformFunctionWrapper.h: Renamed from Source/WebCore/style/values/transforms/StyleTransformOperationWrapper.h.
* Source/WebCore/style/values/transforms/functions/StyleTranslateTransformFunction.cpp: Added.
* Source/WebCore/style/values/transforms/functions/StyleTranslateTransformFunction.h: Added.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAcceleratedEffectStack.mm:
* Source/WebKit/WebProcess/Inspector/WebInspectorBackendClient.cpp:
* Source/WebKit/WebProcess/Inspector/WebInspectorClient.cpp:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDiscretePresentationController.mm:

Canonical link: <a href="https://commits.webkit.org/300845@main">https://commits.webkit.org/300845@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/353d970ed61a85e684b8b1e2ded6f2d06c0b3eae

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124040 "18 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43754 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34452 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130868 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/76190 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/125917 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44479 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52348 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94364 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/76190 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126994 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35444 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110964 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74951 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34391 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/29130 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74351 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105176 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29347 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133539 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50992 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38843 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102827 "9 flakes 57 failures") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51366 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107183 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102641 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26110 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47993 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26245 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/47868 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50845 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56610 "Built successfully") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/50318 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53664 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51992 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/1fc446f5c01408b196c215a10b575bf65fea3b11. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51519

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/3df71feaf6334c3737f4296a714369fea515e2ea. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51519

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/682637c456d245e143d3c5233e8ea0f21fb9516f. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51519

weinig committed:
[Style] Split style and platform transform operation representations
https://bugs.webkit.org/show_bug.cgi?id=299814

Reviewed by Antti Koivisto.

In order to remove another remaining use of WebCore::Length, this change
adds a style specific parallel set of types for representing transform
operations to the WebCore::TransformOperation ones.

The new types, deriving from Style::TransformFunction, use Style primitive
numerics for the underlying data, allowing us to replace the WebCore::Length
with Style::LengthPercentage<> in Style::TranslateTransformFunction.

In the pre-existing WebCore::TranslateTransformOperation, WebCore::Length
is replaced by direct usage of float (it might make sense to change this
to double at some point, but for now, float is used to maintain the same
bit accuracy as Length).

When going from the style types to the platform types, via Style::toPlatform,
a FloatSize must be passed in so that Style::LengthPercentage<> in
Style::TranslateTransformFunction can be resolved down to float.

The platform transform types are only used for platform layer animation
needs (both GraphicsLayer and threaded), so going forward, any capabilities
they don't need can stripped from them. For GraphicsLayer, it may make sense
to remove use of WebCore::TransformOperation entirely, instead having the
ability to go straight from Style::TransformFunction to platform animation
types, be it PlatformCAAnimation or TextureMapperAnimation.

This introduces a bit of duplication of logic for the blending that is not
ideal, so a followup change will aim to add a generic implementation of
the blending algorithms that can be shared.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/BlendingKeyframes.cpp:
* Source/WebCore/animation/KeyframeEffect.cpp:
* Source/WebCore/css/values/CSSValueAggregates.h:
* Source/WebCore/platform/animation/AcceleratedEffect.cpp:
* Source/WebCore/platform/animation/AcceleratedEffect.h:
* Source/WebCore/platform/animation/AcceleratedEffectValues.cpp:
* Source/WebCore/platform/graphics/GraphicsLayer.h:
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/texmap/GraphicsLayerTextureMapper.cpp:
* Source/WebCore/platform/graphics/texmap/GraphicsLayerTextureMapper.h:
* Source/WebCore/platform/graphics/texmap/TextureMapperAnimation.cpp:
* Source/WebCore/platform/graphics/texmap/TextureMapperAnimation.h:
* Source/WebCore/platform/graphics/texmap/coordinated/GraphicsLayerCoordinated.cpp:
* Source/WebCore/platform/graphics/texmap/coordinated/GraphicsLayerCoordinated.h:
* Source/WebCore/platform/graphics/transforms/IdentityTransformOperation.h:
* Source/WebCore/platform/graphics/transforms/Matrix3DTransformOperation.cpp:
* Source/WebCore/platform/graphics/transforms/Matrix3DTransformOperation.h:
* Source/WebCore/platform/graphics/transforms/MatrixTransformOperation.h:
* Source/WebCore/platform/graphics/transforms/PerspectiveTransformOperation.h:
* Source/WebCore/platform/graphics/transforms/RotateTransformOperation.h:
* Source/WebCore/platform/graphics/transforms/ScaleTransformOperation.h:
* Source/WebCore/platform/graphics/transforms/SkewTransformOperation.h:
* Source/WebCore/platform/graphics/transforms/TransformOperation.h:
* Source/WebCore/platform/graphics/transforms/TransformOperations.cpp:
* Source/WebCore/platform/graphics/transforms/TransformOperations.h:
* Source/WebCore/platform/graphics/transforms/TransformOperationsSharedPrimitivesPrefix.h:
* Source/WebCore/platform/graphics/transforms/TranslateTransformOperation.cpp:
* Source/WebCore/platform/graphics/transforms/TranslateTransformOperation.h:
* Source/WebCore/rendering/RenderLayer.cpp:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
* Source/WebCore/rendering/TextAutoSizing.cpp:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+CSSValueConversion.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+CSSValueCreation.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Conversions.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Serialization.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes.h:
* Source/WebCore/style/values/transforms/StyleRotate.cpp:
* Source/WebCore/style/values/transforms/StyleRotate.h:
* Source/WebCore/style/values/transforms/StyleScale.cpp:
* Source/WebCore/style/values/transforms/StyleScale.h:
* Source/WebCore/style/values/transforms/StyleTransform.cpp:
* Source/WebCore/style/values/transforms/StyleTransform.h:
* Source/WebCore/style/values/transforms/StyleTransformFunction.cpp:
* Source/WebCore/style/values/transforms/StyleTransformFunction.h:
* Source/WebCore/style/values/transforms/StyleTransformList.cpp:
* Source/WebCore/style/values/transforms/StyleTransformList.h:
* Source/WebCore/style/values/transforms/StyleTranslate.cpp:
* Source/WebCore/style/values/transforms/StyleTranslate.h:
* Source/WebCore/style/values/transforms/functions/StyleMatrix3DTransformFunction.cpp: Added.
* Source/WebCore/style/values/transforms/functions/StyleMatrix3DTransformFunction.h: Added.
* Source/WebCore/style/values/transforms/functions/StyleMatrixTransformFunction.cpp: Added.
* Source/WebCore/style/values/transforms/functions/StyleMatrixTransformFunction.h: Added.
* Source/WebCore/style/values/transforms/functions/StylePerspectiveTransformFunction.cpp: Added.
* Source/WebCore/style/values/transforms/functions/StylePerspectiveTransformFunction.h: Added.
* Source/WebCore/style/values/transforms/functions/StyleRotateTransformFunction.cpp: Added.
* Source/WebCore/style/values/transforms/functions/StyleRotateTransformFunction.h: Added.
* Source/WebCore/style/values/transforms/functions/StyleScaleTransformFunction.cpp: Added.
* Source/WebCore/style/values/transforms/functions/StyleScaleTransformFunction.h: Added.
* Source/WebCore/style/values/transforms/functions/StyleSkewTransformFunction.cpp: Added.
* Source/WebCore/style/values/transforms/functions/StyleSkewTransformFunction.h: Added.
* Source/WebCore/style/values/transforms/functions/StyleTransformFunctionBase.cpp: Added.
* Source/WebCore/style/values/transforms/functions/StyleTransformFunctionBase.h: Added.
* Source/WebCore/style/values/transforms/functions/StyleTransformFunctionWrapper.h: Renamed from Source/WebCore/style/values/transforms/StyleTransformOperationWrapper.h.
* Source/WebCore/style/values/transforms/functions/StyleTranslateTransformFunction.cpp: Added.
* Source/WebCore/style/values/transforms/functions/StyleTranslateTransformFunction.h: Added.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAcceleratedEffectStack.mm:
* Source/WebKit/WebProcess/Inspector/WebInspectorBackendClient.cpp:
* Source/WebKit/WebProcess/Inspector/WebInspectorClient.cpp:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDiscretePresentationController.mm:

Canonical link: https://commits.webkit.org/300845@main
